/*  Copyright (C) 2024 Me7c7, Martin.JM

    This file is part of gadgetbridge.

    gadgetbridge is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published
    by the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    gadgetbridge is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>. */
package nodomain.freeyourgadget.gadgetfit.service.devices.huawei.requests;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

import nodomain.freeyourgadget.gadgetfit.devices.huawei.HuaweiPacket;
import nodomain.freeyourgadget.gadgetfit.devices.huawei.packets.Contacts;
import nodomain.freeyourgadget.gadgetfit.service.devices.huawei.HuaweiSupportProvider;

public class GetContactsCount extends Request {
    private static final Logger LOG = LoggerFactory.getLogger(GetContactsCount.class);

    public GetContactsCount(HuaweiSupportProvider support) {
        super(support);
        this.serviceId = Contacts.id;
        this.commandId = Contacts.ContactsCount.id;
    }

    @Override
    protected boolean requestSupported() {
        return supportProvider.getDeviceState().supportsContacts();
    }

    @Override
    protected List<byte[]> createRequest() throws RequestCreationException {
        try {
            return new Contacts.ContactsCount.Request(paramsProvider).serialize();
        } catch (HuaweiPacket.CryptoException e) {
            throw new RequestCreationException(e);
        }
    }

    @Override
    protected void processResponse() throws ResponseParseException {
        LOG.debug("handle contacts count");

        if (!(receivedPacket instanceof Contacts.ContactsCount.Response))
            throw new ResponseTypeMismatchException(receivedPacket, Contacts.ContactsCount.Response.class);

        int count = ((Contacts.ContactsCount.Response) receivedPacket).maxCount;
        this.supportProvider.getDeviceState().saveMaxContactsCount(count);
    }
}
